### DISCUSSION TITLE
Issue with fetch on vercel & netlify

### DISCUSSION THREAD

**Discussion by Ahmadh26** [vercel/next.js#43720] [Help]
I have been experiencing this issue for the past week on vercel and netlify

please note that my nextJS version is **12.2.5**

I have the following to fetch data: 
```ts
export default async function fetcher(
  url: string,
  req?: IncomingMessage | undefined,
  method: string = 'GET',
  data?: {}
): Promise<Response> {
  try {
    const baseURl = process.env.NEXT_PUBLIC_BASE_URL
    const cookies = req?.headers
    const res = await fetch(`${baseURl}${url}`, {
      method: method,
      headers: cookies as HeadersInit,
      body: JSON.stringify(data),
    })

    console.log('FETCH RESULT', res.status, res.statusText)
    console.log('FETCH URL', baseURl, url)
    return res
  } catch (error) {
    console.log('FETCH ERROR', error)
    throw error
  }
}
```

this was working fine until recently i started getting a 502 bad gateway error on both vercel and netlify. 

for example (netlify) i get the following: 

```bash
 INFO   FETCH URL https://myWebsiteUrlHere.netlify.app/api /v1/categories/popular
 INFO   FETCH RESULT 502 Bad Gateway
```
yes i know there is a space between `api` and `/v1` that is because of the console log i split them up (check the code up)

now the weird thing is when I open the API url in the browser it returns data as expected. it's just fetch not working properly.

what could be causing this issue? I'm a bit lost rn..

**Edit:** everything is working fine locally (whether it's on dev or prod).

**Comment by icyJoseph** [comment]
So the Netlify app is also a next.js app? Are these two next.js servers talking to each other?

**Reply by Ahmadh26** [reply]
@icyJoseph Not at all. they're both the same app running on 2 different platforms.. i just wanted to see if there was a problem so i deployed on both platforms.

I just realised that this issue is only happening when i have a fetch request in GetServersideProps.. I migrated a few pages to use `useQuery`  in SSR and they started to work normally.. not sure what is happening if you could explain it to me?

**Reply by icyJoseph** [reply]
How often do you see this issue?

So from `GetServerSideProps` (GSSP), you reach out to an `/api` endpoint?

**Reply by Ahmadh26** [reply]
Yes, I reach to an `/api` endpoint and get a 502 error

It has been happening the past few days (probably week)

**Reply by icyJoseph** [reply]
I see, well the GSSP and API route combo is already an anti-pattern. 

https://nextjs.org/docs/basic-features/data-fetching/get-server-side-props#getserversideprops-or-api-routes

> It can be tempting to reach for an [API Route](https://nextjs.org/docs/api-routes/introduction) when you want to fetch data from the server, then call that API route from getServerSideProps. This is an unnecessary and inefficient approach, as it will cause an extra request to be made due to both getServerSideProps and API Routes running on the server.

So it essentially means that from GSSP, you initiate and fly a new request, which is problematic because your API route might as well be an entire different application, and in fact, depending on how the application is deployed, you might hit a CDN, load balancers and what not before the request makes it to the API route.

As the docs say, this is an unnecessary and inefficient approach. You are encouraged to do whatever work you need to do from within GSSP.

Yet! This is probably not the source of the issue, though we ought to keep an eye out for it. 

Is it possible for you to debug what arrives to the API route? Could we somehow confirm that the requests are arriving to the handler?
